File: app/src/main/java/com/example/navigationdrawerapp/DrawerAdapter.java
package com.example.navigationdrawerapp;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

public class DrawerAdapter extends RecyclerView.Adapter<BaseViewHolder> {

    private final String TAG = DrawerAdapter.class.getSimpleName();
    private static final int HeaderView = 0;
    private static final int ContentView = 1;
    private static final int CollapseView = 2;

    private final DrawerListener listener;
    private final List<DrawerModel> list = new ArrayList<DrawerModel>();

    public DrawerAdapter(DrawerListener listener) {
        super();
        this.listener = listener;
    }

    @NonNull
    @Override
    public BaseViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        final LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        switch(viewType) {
            case HeaderView:
                return new HeaderViewHolder (
                    DataBindingUtil.inflate(layoutInflater,
                    R.layout.cell_header, parent, false),
                    listener
                );
            case ContentView:
                return new ContentViewHolder (
                    DataBindingUtil.inflate(layoutInflater, R.layout.cell_content, parent, false),
                    listener
                );
            case CollapseView:
                return new CollapseViewHolder (
                    DataBindingUtil.inflate(layoutInflater, R.layout.cell_collapse, parent, false),
                    listener
                );
            default:
                return new CollapseViewHolder(
                    DataBindingUtil.inflate(layoutInflater, R.layout.cell_collapse, parent, false),
                    listener
                );
            }
    }

    @Override
    public void onBindViewHolder(@NonNull BaseViewHolder holder, int position) {
        holder.bindDataToViewHolder(list.get(position), position);
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    @Override
    public int getItemViewType(int position) {
        if (list.get(position).getHeader()) {
            return HeaderView;
        } else if (!list.get(position).getHeader() && list.get(position).getExpand()) {
            return ContentView;
        } else if (!list.get(position).getHeader() && !list.get(position).getExpand()) {
            return CollapseView;
        } else {
            return super.getItemViewType(position);
        }
    }

    public void setItems(List<DrawerModel> items) {
        list.clear();
        list.addAll(items);
        notifyDataSetChanged();
    }

    public DrawerModel getItem(int position) {
        return list.get(position);
    }

    @Override
    public long getItemId(int position) {
        return super.getItemId(position);
    }

    public void setExpand(DrawerModel model, int position) {
        for (int index = position; index < getItemCount(); index++) {
            if (getItem(index).getHeader() && index != position) {
                break;
            } else if (!getItem(index).getHeader()) {
                getItem(index).setExpand(true);
                notifyItemChanged(index);
            } else {
                getItem(index).setExpand(true);
            }
        }
        //notifyItemRangeChanged(position, getItemCount())
    }

    public void setCompress(DrawerModel model, int position) {
        for (int index = position; index < getItemCount(); index++) {
            if (getItem(index).getHeader() && index != position) {
                break;
            } else if (!getItem(index).getHeader()) {
                getItem(index).setExpand(false);
                notifyItemChanged(index);
            } else {
                getItem(index).setExpand(false);
            }
        }
        //notifyItemRangeChanged(position, getItemCount())
    }
}

File: app/src/main/java/com/example/navigationdrawerapp/DrawerAdapter.kt
package com.example.navigationdrawerapp

import android.util.Log
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.example.navigationdrawerapp.databinding.CollapseBinder
import com.example.navigationdrawerapp.databinding.ContentBinder
import com.example.navigationdrawerapp.databinding.HeaderBinder

class DrawerAdapter : RecyclerView.Adapter<BaseViewHolder> {

    companion object {
        private val TAG = DrawerAdapter::class.java.getSimpleName()
        private const val HeaderView = 0
        private const val ContentView = 1
        private const val CollapseView = 3
    }

    private val listener : DrawerListener
    private val list : MutableList<DrawerModel> = mutableListOf<DrawerModel>()

    constructor(listener : DrawerListener) : super() {
        this.listener = listener
    }

    override fun onCreateViewHolder(parent : ViewGroup, viewType : Int) : BaseViewHolder {
        val layoutInflater : LayoutInflater = LayoutInflater.from(parent.getContext())
        return when(viewType) {
            HeaderView -> {
                val binder : HeaderBinder = DataBindingUtil.inflate(layoutInflater, R.layout.cell_header, parent, false)
                HeaderViewHolder(binder, listener)
            }
            ContentView -> {
                val binder : ContentBinder = DataBindingUtil.inflate(layoutInflater, R.layout.cell_content, parent, false)
                ContentViewHolder(binder, listener)
            }
            CollapseView -> {
                val binder : CollapseBinder = DataBindingUtil.inflate(layoutInflater, R.layout.cell_collapse, parent, false)
                CollapseViewHolder(binder, listener)
            }
            else -> {
                val binder : CollapseBinder = DataBindingUtil.inflate(layoutInflater, R.layout.cell_collapse, parent, false)
                CollapseViewHolder(binder, listener)
            }
        }
    }

    override fun onBindViewHolder(holder : BaseViewHolder, position : Int) {
        holder.bindDataToViewHolder(list.get(position), position)
    }

    override fun getItemCount() : Int {
        return list.size
    }

    override fun getItemViewType(position : Int) : Int {
        return if (list[position].isHeader) {
            HeaderView
        } else if (!list[position].isHeader && list[position].isExpand) {
            ContentView
        } else if (!list[position].isHeader && !list[position].isExpand) {
            CollapseView
        } else {
            super.getItemViewType(position)
        }
    }

    public fun setItems(items : List<DrawerModel>) {
        list.clear()
        list.addAll(items)
        notifyDataSetChanged()
    }

    public fun getItem(position : Int) : DrawerModel {
        return list.get(position)
    }

    override fun getItemId(position : Int) : Long {
        return super.getItemId(position)
    }

    public fun setExpand(model : DrawerModel, position : Int) {
        for (index in position ..< getItemCount()) {
            if (getItem(index).isHeader && index != position) {
                break
            } else {
                Log.d(TAG,"setExpand ${getItem(index)}")
                getItem(index).isExpand = true
                notifyItemChanged(index)
            }
        }
        //notifyItemRangeChanged(position, getItemCount())
    }

    public fun setCompress(model : DrawerModel, position : Int) {
        for (index in position ..< getItemCount()) {
            if (getItem(index).isHeader && index != position) {
                break
            } else {
                Log.d(TAG,"setCompress ${getItem(index)}")
                getItem(index).isExpand = false
                notifyItemChanged(index)
            }
        }
        //notifyItemRangeChanged(position, getItemCount())
    }
}